### PR TITLE
PIM-9807

### PR DESCRIPTION
**Pull request by rfauglas**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php [hunk 1]
```diff
@@ -29,13 +30,18 @@ public function __construct(FieldSplitter $fieldSplitter, array $supportedFieldT
      */
     public function convert(array $attributeFieldInfo, $value)
     {
+        $code = $attributeFieldInfo['attribute']->getCode();
+
+        if ($value instanceof \DateTime) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php:35]:
```suggestion
        if ($value instanceof \DateTimeInterface) {
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php [hunk 2]
```diff
@@ -29,13 +30,18 @@ public function __construct(FieldSplitter $fieldSplitter, array $supportedFieldT
...
+            throw new BusinessArrayConversionException("Can not convert cell  {$code} with date format to attribute of type text", "pim_import_export.notification.export.warnings.xlsx_cell_date_to_text_conversion_error", [$code]);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php:36]: I prefer using `\sprintf()` :slightly_smiling_face:

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/translations/messages.en_US.yml [hunk 3]
```diff
@@ -32,6 +32,10 @@ pim_import_export:
             error:   Export <strong>%label%</strong> failed
             warning: Export <strong>%label%</strong> finished with some warnings
             success: Export <strong>%label%</strong> finished
+            warnings:
+                xlsx_cell_conversion_error: Exception while converting column {0} to target PIM product attribute type (check Excel cell format)
```
Comment (mmetayer) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/translations/messages.en_US.yml:36]:
```suggestion
                xlsx_cell_conversion_error: Exception while converting the "{0}" column to target PIM product attribute type (please check Excel cell format)
```

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/translations/messages.en_US.yml [hunk 4]
```diff
@@ -32,6 +32,10 @@ pim_import_export:
...
+                xlsx_cell_date_to_text_conversion_error: Can not convert cell  {0} with date format to attribute of type text
```
Comment (mmetayer) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/translations/messages.en_US.yml:37]:
```suggestion
                xlsx_cell_date_to_text_conversion_error: Can not convert the "{0}" cell with date format to attribute of type text
```

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/translations/messages.en_US.yml (hunk 3 above)
Comment (mmetayer) [reply src/Akeneo/Platform/Bundle/ImportExportBundle/Resources/translations/messages.en_US.yml:36]: also, a named parameter would be preferable IMO (we usually use the `%parameter%` or `{{ parameter }}` notations; here, something like `%invalidColumn%`)

File: src/Akeneo/Tool/Component/Connector/Exception/BusinessArrayConversionException.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+
+namespace Akeneo\Tool\Component\Connector\Exception;
+
+use Throwable;
```
Comment (mmetayer) [inline src/Akeneo/Tool/Component/Connector/Exception/BusinessArrayConversionException.php:6]: unfortunately the coupling detector does not allow importing classes from the global namespace, you'll have to use `\Throwable` in your code

File: src/Akeneo/Tool/Component/Connector/spec/Reader/File/Xlsx/ReaderSpec.php [hunk 6]
```diff
@@ -19,7 +21,8 @@ function let(
         FileIteratorFactory $fileIteratorFactory,
         ArrayConverterInterface $converter,
         StepExecution $stepExecution
-    ) {
+    )
```
Comment (mmetayer) [inline src/Akeneo/Tool/Component/Connector/spec/Reader/File/Xlsx/ReaderSpec.php:24]: not sure this will pass the `php-cs-fixer`  check

**Review by mmetayer — APPROVED** [review]
GTM once comments are fixed, nice job

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php (hunk 2 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php:36]: Any valid reason? I find it more readable this way....

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php [hunk 7]
```diff
@@ -127,10 +133,10 @@ public function convert(array $values, array $options = [])
     /**
      * @param string $column
      * @param string $value
-     *
+     *q
```
[the remark this answers is not in the document]
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php:136]: bien vu!

File: src/Akeneo/Tool/Component/Connector/spec/Reader/File/Xlsx/ReaderSpec.php (hunk 6 above)
Comment (rfauglas) [reply src/Akeneo/Tool/Component/Connector/spec/Reader/File/Xlsx/ReaderSpec.php:24]: i ran a "fix" before push...

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php (hunk 2 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php:36]: nope, just personal preference, your call :wink:

Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/ValueConverter/TextConverter.php:36]: C/C++ good old times  ;-)

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php [hunk 8]
```diff
@@ -148,7 +149,12 @@ protected function convertValue($column, $value)
                 );
             }
 
-            return $converter->convert($attributeFieldInfo, $value);
+            try {
+                return $converter->convert($attributeFieldInfo, $value);
+            } catch (\Error $e) {
+                throw new BusinessArrayConversionException("Exception while converting column \"{$column}\": bad input format.",
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php:155]: It's not necessary a XLSX file here CSV import also pass throught this code

**Review by StevenVAIDIE — COMMENTED** [review]
For me this bug can be fixed by just adding the following line in Akeneo/Pim/Enrichment/Bundle/Resources/config/readers.yml :

            - {reader_options: {shouldFormatDates: true}}

in line 44, 52, 59, and 109. We don't need that Spout/Reader transform string date to date object. For me the user will not understand the warning generated

cf: http://opensource.box.com/spout/docs/#datetime-formatting

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php (hunk 8 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php:155]: Thanks for the feedback, I didn't know the trick, but it doesn't change the problematic as described  in our exchange in the slack channel (unfortunately i did not report it to jira issue) : there is no locale in the context of the job, so you can not anticipate the expected date formatting locale from the user...

**Comment by mmetayer** [comment]
> For me this bug can be fixed by just adding the following line in Akeneo/Pim/Enrichment/Bundle/Resources/config/readers.yml :
> 
> ```
>         - {reader_options: {shouldFormatDates: true}}
> ```
> 
> in line 44, 52, 59, and 109. We don't need that Spout/Reader transform string date to date object. For me the user will not understand the warning generated

@StevenVAIDIE I had this idea but unfortunately it would break the import for the "date" attributes: if an Excel cell data is a "date", Excel will format it depending on the OS locale (e.g: "dd/mm/yy"); if you the try to import the file with the `shouldFormatDates` set to `true`, there would be a bad format exception, as we only accept 'yyyy-mm-dd' format for date attributes if the data is a string

**Review by anaisbaunelemaire — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php (hunk 8 above)
Comment (rfauglas) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ArrayConverter/FlatToStandard/Value.php:155]: @StevenVAIDIE  are you ok with my answer?

**Pull request merged**